src/expr/src/scalar/func.rs

```diff
@@ -1244,6 +1244,60 @@ fn cot<'a>(a: Datum<'a>) -> Result<Datum, EvalError> {
     Ok(Datum::from(1.0 / f.tan()))
 }
 
+fn log_guard(val: f64, function_name: &str) -> Result<f64, EvalError> {
+    if val.is_sign_negative() {
+        return Err(EvalError::NegativeOutOfDomain(function_name.to_owned()));
+    }
+    if val == 0.0 {
+        return Err(EvalError::ZeroOutOfDomain(function_name.to_owned()));
+    }
+    Ok(val)
+}
+
+fn log_base<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
+    let base = log_guard(b.unwrap_float64(), "log")?;
+    log(a, |f| f.log(base), "log")
+}
+
+fn log_dec<'a, 'b, F: Fn(f64) -> f64>(
+    a: Datum<'a>,
+    logic: F,
+    name: &'b str,
+    scale: u8,
+) -> Result<Datum<'a>, EvalError> {
+    let significand = cast_significand_to_float64(a);
+    let scale = i32::from(scale);
+    let a = log_guard(significand.unwrap_float64() / 10_f64.powi(scale), name)?;
+    Ok(cast_float64_to_decimal(
+        Datum::from(logic(a)),
+        Datum::from(scale),
+    )?)
+}
+
+fn log<'a, 'b, F: Fn(f64) -> f64>(
+    a: Datum<'a>,
+    logic: F,
+    name: &'b str,
+) -> Result<Datum<'a>, EvalError> {
+    let f = log_guard(a.unwrap_float64(), name)?;
+    Ok(Datum::from(logic(f)))
+}
+
+fn exp<'a>(a: Datum<'a>) -> Result<Datum<'a>, EvalError> {
+    Ok(Datum::from(a.unwrap_float64().exp()))
+}
+
+fn exp_dec<'a>(a: Datum<'a>, scale: u8) -> Result<Datum<'a>, EvalError> {
+    let significand = cast_significand_to_float64(a);
```

Comment:
we are following the existing style of the code here so not a huge deal but I'm a little bit weirded out by how many datum conversions we are doing here -- like if I'm reading it right:

line 1291 unwraps a Datum::Decimal and converts it to a Datum::Float64
line 1293 unwraps the Datum::Float64 into a regular f64
lines 1296 and 1297 create two new datums a Datum::Float64 and a Datum::Int32, and then finally
line 1295 converts those into a Datum::Decimal

I didn't look at the generated assembly / don't think this is a big performance deal at this point but I feel like this all makes the code challenging to understand for a first time user. not totally sure why we write these functions like this my guess is that we already have these Datum::Type1 -> Datum::Type2 casts for the actual casting and then its convenient to reuse them here? is there another reason why we are doing things this way?

Reply:
@antifuchs this is me just musing about the code / trying to understand it better. I think the code is all correct as written and we should merge it like this once it passes tests

Reply:
So, I think you're right in feeling weird about this. As we see in the failing test, there's something sketch happening in decimals, and it may very well be rooted in the type conversions. I'll see if behavior changes if we change the type conversion behavior a bit.